### DISCUSSION TITLE
MvvmDialogs with Avalonia 11

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#9014] [Q&A]
I would like some user feedback in regards to [MvvmDialogs for Avalonia](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs).

We got the benefit that Avalonia now supports mobile; and the problem that the OpenFile/SaveFile API is radically different to support mobile devices and browser.

The goal of MvvmDialogs is to create an abstraction layer so that your ViewModel doesn't care whether the UI is WPF, WinUI, Blazor or Avalonia, and allowing you to write the ViewModel without needing any reference to any particular UI classes.

In that sense, Avalonia shares the same goal with its new API.

One option would be to mirror Avalonia's new API into MvvmDialogs.

However...

1. IStorageFile API is actually very complex! It would be a LOT of complex code to duplicate.
2. That complex API is flagged as unstable, and can change at any moment.
3. This more complex API would also affect WPF users who don't need it.

For 3, I'm already providing a WPF-specific sync-methods API that can keep returning string paths instead of IStorageFile, so that may be fine if only the Async methods adopt the new API.

In that case, WPF users wouldn't be too affected whereas Avalonia users do expect the change and need to adapt no matter what to v11.

Is there any better option than mirroring a complex unstable API?

My goals are:
- Supporting Avalonia 11's new OpenFile/SaveFile API
- Potentially supporting MAUI, Blazor and Xamarin Forms if someone decides to implement that in the future
- Provide a STABLE api...

Got feedback before I go break things down?

As I'm looking at the API, it's actually a consumption-only API, so even if details get added or changed, it generally shouldn't affect users. I just need to implement `IStorageBookmarkFile` and `IStorageBookmarkFolder` and expose that.

There is one location where `IStorageFile` is not consumption-only: `PickerOptions.SuggestedStartLocation`.

`BclStorageFile` constructor takes either string or FileInfo in its constructor.

How is that supposed to be set on mobile devices? Does the constructor take what `IStorageItem.TryGetUri` would return?

**Comment by kekekeks** [comment]
`BclStorageFile` class is supposed to be an implementation helper for platforms where we have full file system access, so we don't need to copy-paste code between desktop platforms.

**Comment by maxkatz6** [comment]
Hi!
This API is a new one and we have time to improve it or change it. So thank you for the feedback!

I will try to answer some of your questions/points.

> That complex API is flagged as unstable, and can change at any moment.

We flag all of our platform specific interfaces as unstable, as it can be extended or updated any time.
Also keep in mind, we now [have Unstable and NotClientImplementable attributes](https://github.com/AvaloniaUI/Avalonia/pull/8049). New API uses NotClientImplementable, which means we want to keep it stable but can't guarantee that some members won't be added in the meanwhile. In C# adding a new member to the interface can break compatibility.

> IStorageFile API is actually very complex! It would be a LOT of complex code to duplicate.

I can agree. But we can't just return a string. Ideally I would return something like FileInfo, but .NET API isn't well placed here either.

You can't use "string" (like a path string). You at least need to pass a Uri object. Why? Normally all files would have "file://" schema. But on android you should expect "content://" schema as it returns you a virtual path to the file. You can get a real path most of the time, but it would require additional permissions from user to ask and not really necessary if you work with Android API directly. And browser API returns only relative file paths.

Also, you can't just save a "string" path (or even URI most of the time) to reuse it later. Most of the OS - iOS, macOS, android, browser - require developers to use their bookmark APIs. For instance, after user selected a file or folder, developer need to take a bookmark from it and save the key to the storage. This way application will need to ask for permissions and file picker only once. 
Technically, Windows also has a bookmarks mechanism/sandboxing. But it seemed to die with the UWP.

That's a two problem what FileInfo doesn't solve.
It also mostly has a sync API. You cannot block-wait an async call on the Browser (it's single threaded).

> There is one location where IStorageFile is not consumption-only: PickerOptions.SuggestedStartLocation.

Normally, developer would ask a user to pick a folder (music folder, app folder...) which can be bookmarked and later used as a start location. BclStorageFile would allow a developer to bypass it on some platforms, but it's rather an ugly hack.

Instead I want to extend current API with these methods:
```diff
+ Task<IStorageFile?> TryGetFileFromUri(Uri fileUri);
+ Task<IStorageFolder?> TryGetFolderFromUri(Uri folderUri);
+ Task<IStorageFolder?> TryGetWellKnownFolder(WellKnownFolder);
```

Where the first one is questionable and has little use as for me, other two can be really useful for the file picker API.
So instead of asked user to pick the documents folder or pictures, developer would have direct access to it.
Well known folder would be one of "desktop", "documents", "downloads", "music", "pictures", or "videos" (mirroring of browser API basically).
But there is a problem though. On browser if developer would use TryGetWellKnownFolder, they won't be able to access IStorageFolder properties nor files inside without a picker, which would be a weak point of this API.

**Reply by mysteryx93** [reply]
Not too familiar with the concept of bookmark... let's take a practical case in my app. YouTube Downloader, I want to select the output folder. I display it (perhaps the textbox becomes read-only for cross-platform), store it in a config file, and need to be able to restore that bookmark from a file. I'm not going to preserve the IStorageFolder instance. How is that going to work?

**Reply by maxkatz6** [reply]
About bookmarks:
1. Youtube app asks user to choose folder
2. App uses this folder to save the file
3. Also app bookmarks this folder, gets the string key and saves it somewhere (preferably locally, as these bookmarks are bound to the device). 
4. Next time when app is opened, download folder is resolved from the bookmark, so user don't need to chose it again, unless they want to change it.

**Reply by mysteryx93** [reply]
and how can I construct an IStorageFolder from that bookmark key?

**Reply by maxkatz6** [reply]
With OpenFolderBookmarkAsync
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Platform/Storage/IStorageProvider.cs#L55

**Reply by mysteryx93** [reply]
Does this bookmark stuff work in Windows, or need to write 2 separate code paths whenever doing file-related work?

And if splitting into 2 code branches, what's the best property to use to determine whether to access full file system directly or use bookmarks? There is `IStorageItem.CanBookmark`

And -- whatever API I expose, I need to implement it for WPF. Emulate bookmarks or simply not support it? Probably very hard to support without direct OS support.

**Reply by maxkatz6** [reply]
Yeah, that's a good question.
Currently on Win and Lin we just return file path as a bookmark. Basically, emulation of similar behavior.

**Reply by mysteryx93** [reply]
I think that there needs to be a property to tell whether the file system is directly available, or whether we are limited to the bookmark system.

**Comment by mysteryx93** [comment]
You know what? It occurred to me that we're now dealing with 2 separate themes.

`IDialogService` is doing its job of managing dialogs...

But the real problem is that [System.IO.Abstractions.IFileSystem](https://github.com/TestableIO/System.IO.Abstractions) is not suitable to work with mobile devices.

Perhaps the solution is to drop that service and develop a new base IO service to handle bookmark-based file access?

MvvmDialogs would then in fact be 2 separate services: `IDialogService` and `IBookmarkFileSystem`.

@maxkatz6 can you tell me more specifically what won't work with using standard `System.IO` and things to avoid? It seems that we need to replace certain usage patterns.

This library will need to know whether we have direct access to the file system or not... what's the best way to know that?

I do not feel confident in designing the API of this new service when I've never touched mobile development in my life. Perhaps there are similar projects elsewhere that can be used as base?

If I'd want to do this properly, `IBookmarkFileSystem` would be a separate NuGet package as `MvvmDialogs`, as one could want to use just that without MvvmDialogs. But perhaps it's already done in the mobile ecosystem?

[I found this in regards to accessing the filesystem in iOS](https://learn.microsoft.com/en-us/xamarin/ios/app-fundamentals/file-system). There is no mention of bookmarks, so this looks like one method among others. There are various platform-specific things to take into consideration.

**Comment by kekekeks** [comment]
> There is no mention of bookmarks, so this looks like one method among others. There are various platform-specific things to take into consideration.

There is full access to you app's own directory. You can manipulate files in your data dir using System.IO. The main problem is files in locations that are shared between multiple apps.

**Comment by mysteryx93** [comment]
OK so it would be an additional service to access files in a different way.

Methods would simply be like:
- GetFileFromBookmark
- GetFileFromPath
- GetFolderFromBookmark
- GetFolderFromPath
- DeleteBookmark (?)
- FileSystemAccess (Full, Container, etc)

This would return `IDialogStorateFile`/`IDialogStorageFolder`. Doesn't really go well directly on `IDialogService` so probably better to have separate service `IBookmarkFileSystem`.

FileSystemAccess, what would be the list of access types available?

To delete bookmarks, where would be the best place to do that?

Anything else missing?

**Comment by mysteryx93** [comment]
I'm breaking my head over mobile navigation support... Just to give an idea of where I'm at.

I implement a `NavigationManager` and a `NavigationRoot` user control when running a single-view app. `NavigationManager` contains a navigation history, and responds to `Show` and `ShowDialog` commands.

When building a history, I'm better to store only the ViewModel to save memory. The View type can be determined from the ViewModel type, and the View can be reconstructed from the ViewModel.

The difference between `Show` and `ShowDialogAsync` is that `ShowDialogAsync` blocks the calling code until it returns a value. Well -- that means that the parent UserControl won't get unloaded, and when I want to navigate back to the parent, I'm going to reconstruct a new View on top of the existing one?

Arggg..... really need to think this one through.

Eventually, if the mobile back button can be accessed via API, perhaps the `MvvmDialogs` library can handle it automatically. That would be cool.

Adding mobile support is quite a lot of work and hassles...
